.github/workflows/deploy.yml

```diff
@@ -0,0 +1,97 @@
+name: Deploy
+
+on:
+  workflow_dispatch:
+    inputs:
+      environment:
+        description: 'Deploy environment ( dev, test, preprod or production )'
+        required: true
+        default: dev
+        type: environment
+        options:
+          - dev
+          - test
+          - preprod
+          - production
+  push:
+    branches:
+      - main
+env:
+  CONTAINER_REGISTRY: ghcr.io
+
+jobs:
+  docker:
+    name: Docker build and push
+    runs-on: ubuntu-latest
+    outputs:
+      docker_image: ${{ steps.image.outputs.tag }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Docker image tag
+        id: image
+        run: |
+          echo ::set-output name=tag::$CONTAINER_REGISTRY/$(echo $GITHUB_REPOSITORY | tr '[:upper:]' '[:lower:]'):$GITHUB_SHA
+
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ${{ env.CONTAINER_REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Docker build & push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          push: true
+          tags: ${{ steps.image.outputs.tag }}
+
+  deploy_dev:
+    name: Deploy to dev environment
+    needs: [docker]
+    runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main' && github.event_name == 'push'
+    environment:
+      name: dev
+      url: ${{ steps.deploy.outputs.environment_url }}
+    concurrency: deploy_dev
+
+    outputs:
+      environment_url: ${{ steps.deploy.outputs.environment_url }}
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/workflows/actions/deploy-environment
+        id: deploy
+        with:
+          environment_name: dev
+          docker_image: ${{ needs.docker.outputs.docker_image }}
+          azure_credentials: ${{ secrets.AZURE_CREDENTIALS }}
+          terraform_vars: workspace_variables/dev.tfvars.json
+
+  deploy_environment:
```

Comment:
this job needs to be continuous deployment for other envs. Maybe ignore for now and pick it up after dev environment is deployed?